internal/snippets/snippets.go

```diff
@@ -0,0 +1,279 @@
+	s := sm.protoServices[serviceName]
+	return fmt.Sprintf("%s_%s_generated_%s_%s_sync", s.shortName, sm.apiVersion, serviceName, methodName)
+}
+
+// ToMetadataJSON marshals the completed SnippetMetadata to a []byte containing
+// the protojson output.
+func (sm *SnippetMetadata) ToMetadataJSON() ([]byte, error) {
+	m := sm.ToMetadataIndex()
+	b, err := protojson.MarshalOptions{Multiline: true}.Marshal(m)
+	if err != nil {
+		return nil, err
+	}
+	// Hack to standardize output from protojson which is currently
+	// non-deterministic with spacing after json keys.
+	return spaceSanitizerRegex.ReplaceAll(b, []byte(": ")), nil
+}
+
+// ToMetadataIndex creates a metadata.Index from the SnippetMetadata.
+func (sm *SnippetMetadata) ToMetadataIndex() *metadata.Index {
+	index := &metadata.Index{
+		ClientLibrary: &metadata.ClientLibrary{
+			Name:     sm.libPkg,
+			Version:  VersionPlaceholder,
+			Language: metadata.Language_GO,
+			Apis: []*metadata.Api{
+				{
+					Id:      sm.protoPkg,
+					Version: sm.protoVersion(),
+				},
+			},
+		},
+	}
+
+	// Sort keys to stabilize output
+	var svcKeys []string
+	for k := range sm.protoServices {
+		svcKeys = append(svcKeys, k)
+	}
+	sort.StringSlice(svcKeys).Sort()
+	for _, serviceShortName := range svcKeys {
+		reducedServName := pbinfo.ReduceServName(serviceShortName, sm.pkgName)
+		clientShortName := reducedServName + "Client"
+		service := sm.protoServices[serviceShortName]
+		var methodKeys []string
+		for k := range service.methods {
+			methodKeys = append(methodKeys, k)
+		}
+		sort.StringSlice(methodKeys).Sort()
+		for _, methodShortName := range methodKeys {
+			method := service.methods[methodShortName]
+			snp := &metadata.Snippet{
+				RegionTag:   method.regionTag,
+				Title:       fmt.Sprintf("%s %s Sample", service.shortName, methodShortName),
+				Description: strings.TrimSpace(method.doc),
+				File:        fmt.Sprintf("%s/%s/main.go", clientShortName, methodShortName),
+				Language:    metadata.Language_GO,
+				Canonical:   false,
+				Origin:      *metadata.Snippet_API_DEFINITION.Enum(),
+				ClientMethod: &metadata.ClientMethod{
+					ShortName:  methodShortName,
+					FullName:   fmt.Sprintf("%s.%s.%s", sm.protoPkg, clientShortName, methodShortName),
+					Async:      false,
+					ResultType: method.result,
+					Client: &metadata.ServiceClient{
+						ShortName: clientShortName,
+						FullName:  fmt.Sprintf("%s.%s", sm.protoPkg, clientShortName),
+					},
+					Method: &metadata.Method{
+						ShortName: methodShortName,
+						FullName:  fmt.Sprintf("%s.%s.%s", sm.protoPkg, service.protoName, methodShortName),
+						Service: &metadata.Service{
+							ShortName: service.protoName,
+							FullName:  fmt.Sprintf("%s.%s", sm.protoPkg, service.protoName),
+						},
+					},
+				},
+			}
+			segment := &metadata.Snippet_Segment{
+				Start: int32(method.regionTagStart + 1),
+				End:   int32(method.regionTagEnd - 1),
```

Comment:
Could you add a comment explaining why we must +/- 1 on these values? Just for future reference.

Reply:
Done, thanks!